torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -350,6 +350,13 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeStreams() {
   for (size_t i = 0; i < devices_.size(); ++i) {
     auto currentStream = at::cuda::getCurrentCUDAStream(devices_[i].index());
     // Block the current stream on the NCCL stream
+    // Check to ensure event is created. Otherwise, this block() call will be a
+    // no-op which can cause silent correctness issues as there will be no
+    // stream synchronization.
+    TORCH_CHECK(
+      (*cudaEvents_)[i].isCreated(),
+      c10::str("Invalid CUDA event on device ", i)
+    );
     (*cudaEvents_)[i].block(currentStream);
```

Comment:
There are quite a few other places where we use this API:

```
grep -R 'block(' torch/lib/*
torch/lib/c10d/ProcessGroupNCCL.cpp:    ncclEvent.block(ncclStream);
torch/lib/c10d/ProcessGroupNCCL.cpp:    (*cudaEvents_)[i].block(currentStream);
torch/lib/c10d/ProcessGroupGloo.cpp:    events[i].block(streams[i]);
torch/lib/c10d/ProcessGroupGloo.cpp:    events[i].block(streams[i]);
torch/lib/c10d/ProcessGroupGloo.cpp:      events[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:      events[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:      events[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:      events[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:      outputEvents[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:      outputEvents[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:      outputEvents[i].block(at::cuda::getCurrentCUDAStream());
torch/lib/c10d/ProcessGroupGloo.cpp:    outputEvents.front().block(at::cuda::getCurrentCUDAStream());
```

Maybe we should update `CUDAEvent.block` to assert that the event is created? If that might break existing code, maybe have an optional flag in `CUDAEvent.block` which when set to `true`, causes `block` to fail?

cc @mruberry

Reply:
Sounds good, we should probably add it at the lower-level `block()` API so that we can cover all these use cases. @mruberry Would there be any issues in doing so?

Reply:
I wish I could give you a definitive answer @rohan-varma. We're talking about changing this line, right?

https://github.com/pytorch/pytorch/blob/48b6b9221a544f7d5b7d6f93eec2fa304da8af94/aten/src/ATen/cuda/CUDAEvent.h#L123

Conceptually I think that's OK to do, although I don't recall if there was a reason for the approach the code currently takes. Since the API is internal an audit of callsites (like the above) should help determine if the behavior is relied upon.

Note that adding an assertion may also impact performance.

Reply:
@mruberry Thanks, that is the line we are looking to change. Since this could have a large blast and be performance sensitive, I think it's probably best to take @pritamdamania87 's suggestion to enable this only with a flag that we pass in at the c10d callsites?

Reply:
That would reduce the risk, yep. There are several reasonable implementations but you might just want:

```
void block_or_assert(const CUDAStream& stream) {
  TORCH_INTERNAL_ASSERT(is_created_);
  CUDAGuard guard(stream.device_index());
  AT_CUDA_CHECK(cudaStreamWaitEvent(stream, event_, 0));
}
```

`TORCH_CHECK` is only for errors that could be produced from user inputs.